### PR TITLE
Adding support for cvc5

### LINKED ISSUE #202: Use `ALL` instead of `ALL_SUPPORTED`?
Currently, `what4`'s [Boolector](https://github.com/GaloisInc/what4/blob/6eda4e395c5fa5110e26ace6b71b11d753788b6f/what4/src/What4/Solver/Boolector.hs#L131) and [CVC4 backends](https://github.com/GaloisInc/what4/blob/719a3b890e661783b504d5380ea39b95d66a9b5c/what4/src/What4/Solver/CVC4.hs#L167) use `ALL_SUPPORTED` as the default logic. I recently did a quick experiment to see if What4's CVC4 settings would work for CVC5, but it turns out that this is not the case, as CVC5 does not recognize `ALL_SUPPORTED`:

```
$ echo "(set-logic ALL_SUPPORTED)" > test.smt2
$ cvc4 test.smt2 
$ cvc5 test.smt2 
(error "Parse Error: test.smt2:1.23: Illegal argument detected

  (set-logic ALL_SUPPORTED)
             ^

void cvc5::internal::LogicInfo::setLogicString(std::string)

  `logicString' is a bad argument
LogicInfo::setLogicString(): junk ("LL_SUPPORTED") at end of logic string: ALL_SUPPORTED")
```

This is a result of cvc5/cvc5#6664, which removes `ALL_SUPPORTED` in favor of `ALL`. Unlike `ALL_SUPPORTED`, `ALL` is actually mentioned by name in the SMT-LIB standard. (See [Section 4.2.1](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2017-07-18.pdf).)

In order to support CVC5, we'll need to add support for `ALL` at the minimum. But I wonder if we shouldn't just use `ALL` everywhere? Being mentioned in the SMT-LIB standard, it's supported by every SMT solver I've tried.

### PR DESCRIPTION
**Pull request by arjunvish**
Adding cvc5 support
Fixes #202

### CODE REVIEW

File: what4/src/What4/Protocol/SMTLib2/Syntax.hs [hunk 1]
```diff
@@ -184,7 +185,11 @@ qf_bv = Logic "QF_BV"
 -- | Set the logic to all supported logics.
 allSupported :: Logic
 allSupported = Logic "ALL_SUPPORTED"
+{-# DEPRECATED allSupported "Use allLogic instead" #-}
```
Comment (RyanGlScott) [inline what4/src/What4/Protocol/SMTLib2/Syntax.hs:188]: It would be worth noting in `what4/CHANGES.md` that `allSupported` is now deprecated in favor `allLogic`.

File: what4/src/What4/Solver/CVC5.hs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+--
+-- CVC5-specific tweaks to the basic SMTLib2 solver interface.
+------------------------------------------------------------------------
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+
+module What4.Solver.CVC5
```
Comment (RyanGlScott) [inline what4/src/What4/Solver/CVC5.hs:18]: It would be worth mentioning in the changelog that `what4` supports CVC5 now.

File: what4/src/What4/Solver/CVC5.hs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+cvc5Path = configOption knownRepr "solver.cvc5.path"
+
+cvc5RandomSeed :: ConfigOption BaseIntegerType
+cvc5RandomSeed = configOption knownRepr "solver.cvc5.random-seed"
+
+-- | Per-check timeout, in milliseconds (zero is none)
+cvc5Timeout :: ConfigOption BaseIntegerType
+cvc5Timeout = configOption knownRepr "solver.cvc5.timeout"
+
+-- | Control strict parsing for Boolector solver responses (defaults
```
Comment (RyanGlScott) [inline what4/src/What4/Solver/CVC5.hs:72]: There's a typo here:

```suggestion
-- | Control strict parsing for CVC5 solver responses (defaults
```

In fact, the same typo is also present in `What4.Solver.CVC4`. Would you mind updating that file as well?

File: what4/test/ExprBuilderSMTLib2.hs [hunk 4]
```diff
@@ -1191,7 +1191,7 @@ testResolveSymBV searchStrat =
 main :: IO ()
 main = do
   testLevel <- TestLevel . fromMaybe "0" <$> lookupEnv "CI_TEST_LEVEL"
-  let solverNames = SolverName <$> [ "cvc4", "yices", "z3" ]
+  let solverNames = SolverName <$> [ "cvc4", "yices", "z3" ] -- TODO RGS: Add cvc5 here?
```
Comment (RyanGlScott) [inline what4/test/ExprBuilderSMTLib2.hs:1194]: Indeed, we might want to follow this TODO's advice and add `cvc5` to this list. I'm not exactly sure if it will work out of the box, but it would be worth a try!

**Comment by arjunvish** [comment]
> 

Not at all, thanks for the comments, they make my work easier :)

**Comment by RyanGlScott** [comment]
The current CI failure is:

```
error: attribute 'cvc5' missing
       at /nix/store/r76czr590rslx3059qrjkgkxbs8741hm-source/flake.nix:180:18:
          179|           };
          180|           cvc5 = pkgs.cvc5 // { # whatever the nixpkgs current version is...
             |                  ^
          181|             v1_0_0  = mkCVC5 "1.0.0";
```

If I had to guess, this is because the CI is using too old of a version of `nixpkgs` to know what `cvc5` is. I think this can be fixed by updating the `nixpkgs` version to a newer one, such as `nixos-22.05`:

https://github.com/GaloisInc/what4/blob/e569514d10f940db20027705388565f3d9f729e3/.github/workflows/test.yml#L42

(@kquick, please correct me if I am wrong here.)

**Comment by kquick** [comment]
That's essentially true, but this will need a two-stage fix.  

First, as @RyanGlScott noted, nixos-22.05 is the first version of nixos that recognizes cvc5.  The nix version that needs to be changed for this is in the GaloisInc/flakes repo.  I believe that switching from the current 21.11 to 22.05 will still work for all older solver versions, so this change should be simple (but this PR will need to use this updated version, which should happen automatically).

Then we have the issue where GHC 8.6.5 was dropped from more recent nixos versions, so we cannot test the combination of GHC 8.6.5 && CVC5.  There are two ways to address this:
 *  The Blue Pill solution is to simply drop GHC 8.6.5 support; that should be reasonable because we have a typical range of 3 GHC versions, and we are ignoring GHC 9.0, but GHC 9.2, GHC 8.10, and GHC 8.8 are probably a reasonable set.  To take this approach simply remove the `version(ghc, '8.6.5').` line from `gen_matrix.pl`.  
 * The Red Pill approach is to setup an exclusion in `gen_matrix.pl` that will cause it to avoid specific combinations.  This is probably something that should be implemented anyhow, but it would involve adding a new prolog fact (e.g. `avoid(ghc, '8.6.5', cvc5, _).`) and then adding a verification that the fact is not present when emitting output.  I'd probably add a `\+ avoid(Compiler, CompilerVersion, Solver, SolverVersion)` clause to `matrix_entry`, but that would involve extracting the latter two from `Solvers`.  This would be useful to have in general if you feel your Prolog skills are up to it, but if that's too much of a distraction, you can just take the Blue Pill.

**Comment by RyanGlScott** [comment]
> Then we have the issue where GHC 8.6.5 was dropped from more recent nixos versions, so we cannot test the combination of GHC 8.6.5 && CVC5.

I'm a bit confused. I thought that the CI was using a _different_ `nixpkgs` version to obtain GHC? Looking at these lines:

https://github.com/GaloisInc/what4/blob/e569514d10f940db20027705388565f3d9f729e3/.github/workflows/test.yml#L97-L104

If I'm reading that correctly, the `nix shell` command is pinning the `nixpkgs` version used for downloading `cabal` + GHC to specific versions, while other tools inherit the default `nixpkgs` version that is provided by `install-nix-action`. Is that not the case?

Assuming that doesn't work, is there any reason why we have to tie the versions of GHC that we use to Nix? Couldn't we just as well download GHC from [`haskell/actions`](https://github.com/haskell/actions) and download the remaining tools from Nix?

**Comment by kquick** [comment]
The snippet you highlighted does cover the GHC, but the solvers are provided via `nix run github:GaloisInc/flakes#cvc5.v${{ matrix.cvc5 }}` which references https://github.com/GaloisInc/flakes/blob/main/flake.nix that contains its own nixpkgs reference.  It's possible to override that with the nixpkgs chosen by this script, but the GaloisInc/flakes should be functional in a standalone manner and that override shouldn't be necessary for an executable.

The nix approach can take better advantage of caching than the haskell/actions approach.  As a rough comparison, the what4 builds take approximately 1m40s to start the actual build step, whereas crucible (excluding additional rust preparation) takes about 4m to reach the same point.  And I'm hoping for even better performance when some pending work lands.

**Comment by RyanGlScott** [comment]
Got it, thanks for the explanation. In that case, I suppose I don't have particularly strong feelings about which of the two options you propose is more desirable. If we went with the approach of excluding CVC5 from GHC 8.6.5 builds, we'd have to add some additional CPP to make sure that the test suite are runnable without CVC5 installed. This would be a bit annoying, but not unprecedented, since we already put similar CPP in place for STP and dReal. See [here](https://github.com/GaloisInc/what4/blob/e569514d10f940db20027705388565f3d9f729e3/what4/what4.cabal#L49-L57).

**Comment by arjunvish** [comment]
It seems to me like neither of you have a preference for one of the options over the other. Since I'm trying to get this integrated as quickly as possible, and since my Prolog skills are pretty poor, I'm leaning on the easy way out, which seems to be option 1. Let me know if either of you have any objections.

File: what4/test/OnlineSolverTest.hs [hunk 5]
```diff
@@ -371,8 +373,9 @@ timeoutTests testLevel solvers =
       -- developer as solvers, What4 formulation, and machine speeds
       -- evolve.
       approxTestTimes :: [ (SolverName, Time) ]
-      approxTestTimes = [ (SolverName "Z3",         2.27 % Second)    -- Z3 4.8.10.  Z3 is good at self timeout.
+      approxTestTimes = [ (SolverName "Z3",         2.7 % Second)    -- Z3 4.8.10.  Z3 is good at self timeout.
```
Comment (RyanGlScott) [inline what4/test/OnlineSolverTest.hs:376]: Why was this changed?

File: what4/src/What4/Solver/CVC5.hs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  smtlib2StructSort tps = Syntax.Sort $ "(Tuple" <> foldMap f tps <> ")"
+    where f x = " " <> Syntax.unSort x
+
+  smtlib2StructCtor args = Syntax.term_app "mkTuple" args
+  smtlib2StructProj _n i x = Syntax.term_app (Syntax.builder_list ["_", "tupSel", fromString (show i)]) [ x ]
+
+cvc5Features :: ProblemFeatures
+cvc5Features = useComputableReals
+           .|. useLinearArithmetic
+           .|. useNonlinearArithmetic
```
Comment (RyanGlScott) [inline what4/src/What4/Solver/CVC5.hs:132]: These two entries are redundant, as they are [implied by `useIntegerArithmetic` and `useComputableReals`](https://github.com/GaloisInc/what4/blob/e569514d10f940db20027705388565f3d9f729e3/what4/src/What4/ProblemFeatures.hs#L70-L76).

**Review by RyanGlScott — COMMENTED** [review]
The CI is running now. Hooray! (One job failed due to the same underlying issues as #174, but don't worry about that too much—that's a transient error that should go away on a subsequent run.)

File: what4/test/OnlineSolverTest.hs (hunk 5 above)
Comment (arjunvish) [reply what4/test/OnlineSolverTest.hs:376]: Good catch, I was playing with the numbers and changed it back to the wrong time.

**Comment by RyanGlScott** [comment]
One remaining mystery: for some reason, it now takes much longer than before to run the test suite with these changes. For instance, it took 2m 34s to run the test suite on `master` (see [here](https://github.com/GaloisInc/what4/runs/7115750979?check_suite_focus=true#step:12:1)), but on this PR, it took 1h 18m 58s (see [here](https://github.com/GaloisInc/what4/runs/7141939577?check_suite_focus=true#step:12:1)). I'm not quite sure what is going on, but we should definitely try to figure out what causes this.

**Comment by RyanGlScott** [comment]
I was mistaken: the CI taking an hour to run the tests is _not_ something that was introduced in this PR. I recently restarted the CI jobs on the latest commit in `master` ([here]()), and I observe the same thing as well. I'm forced to conclude that bumping the `nixpkgs` version in the `flakes` repo to 22.05 (see GaloisInc/flakes@57044d2a7ed55d8e4f74bb7dc2ca8e984c6ecce1) is responsible, but I still don't really have a good sense of why that is happening.

**Comment by RyanGlScott** [comment]
I've opened #207 to discuss the matter of the CI slowdown separately.

**Comment by kquick** [comment]
I made a note on #207, but basically I believe this was just due to cache seeding.  Also of note: builds from this PR will be unable to push to cache since the cache key is a repository secret and this PR comes from a fork, rather than a branch on the main repository.

**Comment by RyanGlScott** [comment]
Thanks! I was starting to worry that `nixpkgs` had undergone some horrible performance regression, but you have brought me back to my senses.

File: what4/test/ExprBuilderSMTLib2.hs [hunk 7]
```diff
@@ -1191,7 +1191,7 @@ testResolveSymBV searchStrat =
 main :: IO ()
 main = do
   testLevel <- TestLevel . fromMaybe "0" <$> lookupEnv "CI_TEST_LEVEL"
-  let solverNames = SolverName <$> [ "cvc4", "yices", "z3" ]
+  let solverNames = SolverName <$> [ "cvc4", "cvc5", "yices", "z3" ]
```
Comment (RyanGlScott) [inline what4/test/ExprBuilderSMTLib2.hs:1194]: While poking around at the test suites recently, I discovered that adding a new solver to the solver tests alone doesn't do much. In order to actually _use_ CVC5 in this test suite, you'd need to add some tests here:

https://github.com/GaloisInc/what4/blob/cff655cce036c70eb113a2f9b1d98dec8c88c2a4/what4/test/ExprBuilderSMTLib2.hs#L1311-L1313

Perhaps it would suffice to define `cvc5Tests = cvc4Tests` for now. If there are CVC5-specific test cases that we encounter later, we can always add on to `cvc5Tests` in the future.

File: what4/test/OnlineSolverTest.hs [hunk 8]
```diff
@@ -373,6 +375,7 @@ timeoutTests testLevel solvers =
       approxTestTimes :: [ (SolverName, Time) ]
       approxTestTimes = [ (SolverName "Z3",         2.27 % Second)    -- Z3 4.8.10.  Z3 is good at self timeout.
                         , (SolverName "CVC4",       7.5  % Second)    -- CVC4 1.8
+                        , (SolverName "CVC5",       0.40  % Second)   -- CVC5 1.0.0
```
[the remark this answers is not in the document]
Comment (arjunvish) [reply what4/test/OnlineSolverTest.hs:378]: Will open an issue for redesigning the challenge problem

**Comment by arjunvish** [comment]
> I made a note on #207, but basically I believe this was just due to cache seeding. Also of note: builds from this PR will be unable to push to cache since the cache key is a repository secret and this PR comes from a fork, rather than a branch on the main repository.

Hm, I was debating whether to push to Ryan's old branch in the repo initially, and admittedly, I didn't think of this consequence. At the very least, I'll keep this in mind for future PRs.

**Review by RyanGlScott — APPROVED** [review]
LGTM!
